### PR TITLE
make installation  error details become more clear when building kubernetes client

### PR DESCRIPTION
**Pull request by zhlsunshine**
make installation  error details become more clear when building kubernetes client based on [issue#33402](https://github.com/istio/istio/issues/33402)

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by zhlsunshine** [comment]
I have done 2 test cases in my environment as below:

**With wrong value of KUBECONFIG:**
```
$ istioctl install --set profile=demo -y
Error: encounter error when creating extended kubernetes client according to given parameters, error detail:(error loading config file "/root": read /root: is a directory)
```

**With wrong information inside config file of KUBECONFIG:**
```
$ istioctl install --set profile=demo -y
Error: encounter error when fetching kubernetes config file and clientset according to given config path, error detail:(Get "https://10.238.154.68:6444/api?timeout=32s": dial tcp 10.238.154.68:6444: connect: connection refused)
```

File: operator/cmd/mesh/install.go [hunk 1]
```diff
@@ -348,3 +348,7 @@ func getProfileNSAndEnabledComponents(iop *v1alpha12.IstioOperator) (string, str
 	}
 	return iop.Spec.Profile, name.IstioDefaultNamespace, enabledComponents, nil
 }
+
+func wrapErrorf(operation string, err error) error {
+	return fmt.Errorf("encounter error when %s, error detail:(%v)", operation, err)
```
Comment (howardjohn) [inline operator/cmd/mesh/install.go:353]:
```suggestion
	return fmt.Errorf("%s: %v", operation, err)
```

File: operator/cmd/mesh/install.go [hunk 2]
```diff
@@ -134,24 +134,24 @@ func runApplyCmd(cmd *cobra.Command, rootArgs *rootArgs, iArgs *installArgs, log
 	var opts clioptions.ControlPlaneOptions
 	kubeClient, err := kube.NewExtendedClient(kube.BuildClientCmd(iArgs.kubeConfigPath, iArgs.context), opts.Revision)
 	if err != nil {
-		return err
+		return wrapErrorf("creating extended kubernetes client according to given parameters", err)
```
Comment (howardjohn) [inline operator/cmd/mesh/install.go:137]:
```suggestion
		return wrapErrorf("creating Kubernetes client", err)
```

File: operator/cmd/mesh/install.go [hunk 3]
```diff
@@ -134,24 +134,24 @@ func runApplyCmd(cmd *cobra.Command, rootArgs *rootArgs, iArgs *installArgs, log
...
 	}
 	restConfig, clientset, client, err := K8sConfig(iArgs.kubeConfigPath, iArgs.context)
 	if err != nil {
-		return err
+		return wrapErrorf("fetching kubernetes config file and clientset according to given config path", err)
```
Comment (howardjohn) [inline operator/cmd/mesh/install.go:141]:
```suggestion
		return wrapErrorf("fetching Kubernetes config file", err)
```

File: operator/cmd/mesh/install.go [hunk 4]
```diff
@@ -134,24 +134,24 @@ func runApplyCmd(cmd *cobra.Command, rootArgs *rootArgs, iArgs *installArgs, log
...
 	}
 	if err := k8sversion.IsK8VersionSupported(clientset, l); err != nil {
-		return err
+		return wrapErrorf("checking minimum supported Kubernetes version for istio", err)
```
Comment (howardjohn) [inline operator/cmd/mesh/install.go:144]:
```suggestion
		return wrapErrorf("checking minimum supported Kubernetes version", err)
```

File: operator/cmd/mesh/install.go [hunk 5]
```diff
@@ -134,24 +134,24 @@ func runApplyCmd(cmd *cobra.Command, rootArgs *rootArgs, iArgs *installArgs, log
...
 	}
 	tag, err := GetTagVersion(operatorVer.OperatorVersionString)
 	if err != nil {
-		return err
+		return wrapErrorf("fetching istio version", err)
```
Comment (howardjohn) [inline operator/cmd/mesh/install.go:148]:
```suggestion
		return wrapErrorf("fetching Istio version", err)
```

File: operator/cmd/mesh/install.go [hunk 6]
```diff
@@ -134,24 +134,24 @@ func runApplyCmd(cmd *cobra.Command, rootArgs *rootArgs, iArgs *installArgs, log
...
 	}
 	setFlags := applyFlagAliases(iArgs.set, iArgs.manifestsPath, iArgs.revision)
 
 	_, iop, err := manifest.GenerateConfig(iArgs.inFilenames, setFlags, iArgs.force, restConfig, l)
 	if err != nil {
-		return err
+		return wrapErrorf("createing an IstioOperatorSpec from different sources", err)
```
Comment (howardjohn) [inline operator/cmd/mesh/install.go:154]:
```suggestion
		return wrapErrorf("generating config", err)
```

File: operator/cmd/mesh/install.go (hunk 1 above)
Comment (zhlsunshine) [reply operator/cmd/mesh/install.go:353]: done!

File: operator/cmd/mesh/install.go (hunk 2 above)
Comment (zhlsunshine) [reply operator/cmd/mesh/install.go:137]: done

File: operator/cmd/mesh/install.go (hunk 3 above)
Comment (zhlsunshine) [reply operator/cmd/mesh/install.go:141]: done

File: operator/cmd/mesh/install.go (hunk 4 above)
Comment (zhlsunshine) [reply operator/cmd/mesh/install.go:144]: done

File: operator/cmd/mesh/install.go (hunk 5 above)
Comment (zhlsunshine) [reply operator/cmd/mesh/install.go:148]: done

File: operator/cmd/mesh/install.go (hunk 6 above)
Comment (zhlsunshine) [reply operator/cmd/mesh/install.go:154]: done

File: operator/cmd/mesh/install.go [hunk 7]
```diff
@@ -134,29 +134,35 @@ func runApplyCmd(cmd *cobra.Command, rootArgs *rootArgs, iArgs *installArgs, log
 	var opts clioptions.ControlPlaneOptions
 	kubeClient, err := kube.NewExtendedClient(kube.BuildClientCmd(iArgs.kubeConfigPath, iArgs.context), opts.Revision)
 	if err != nil {
-		return err
+		operation := "encounter error when creating Kubernetes client, error detail"
+		return wrapErrorf(operation, err)
```
Comment (howardjohn) [inline operator/cmd/mesh/install.go:138]:
```suggestion
		return fmt.Errorf("create Kubernetes client: %v", err)
```

This follows standard patterns in the rest of the code base. When creating errors like this, we get useful messages as errors propogate up, like `installation failed: create Kubernetes client: permission denied` or similar

Comment (howardjohn) [reply operator/cmd/mesh/install.go:138]: similar comment applies to the rest

Comment (zhlsunshine) [reply operator/cmd/mesh/install.go:138]: got it

Comment (zhlsunshine) [reply operator/cmd/mesh/install.go]: @howardjohn the code style is the same now. ^_^

**Review by howardjohn — APPROVED** [review]

**Pull request merged**